behind_ingensec.rb

```diff
@@ -0,0 +1,285 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Behind BinarySec/IngenSecurity',
+      'Version' => '$Release: 1.0.1',
+      'Description'    => %q{
+        This module can be useful if you need to test
+        the security of your server and your website
+        behind BinarySec/IngenSec by discovering the real IP address.
+      },
+      'Author'         => [
+        'mekhalleh',
+        'RAMELLA Sébastien <sebastien.ramella[at]Pirates.RE>'
+      ],
+      'References' => [
+        ['URL', 'https://www.ingensecurity.com/fct_apercu.html']
+      ],
+      'License'        => MSF_LICENSE
+    ))
+
+    register_options(
+      [
+        OptString.new('HOSTNAME', [true, 'The hostname to find real IP', 'www.reunionnaisdumonde.com']),
+        OptString.new('URIPATH', [true, 'The URI path (for custom comparison)', '/']),
+        OptInt.new('RPORT', [true, 'The target port (for custom comparison)', 443]),
+        OptBool.new('SSL', [true, 'Negotiate SSL/TLS for outgoing connections (for custom comparison)', true]),
+        OptString.new('Proxies', [false, 'A proxy chain of format type:host:port[,type:host:port][...]']),
+        OptInt.new('THREADS', [false, 'Threads for DNS enumeration', 1]),
+        OptPath.new('WORDLIST', [true, 'Wordlist of subdomains', ::File.join(Msf::Config.data_directory, 'wordlists', 'namelist.txt')])
+      ])
+  end
+
+  def do_check_tcp_port(ip, port, proxies)
+    begin
+      sock = Rex::Socket::Tcp.create(
+        'PeerHost' => ip,
+        'PeerPort' => port,
+        'Proxies'  => proxies
+      )
+    rescue ::Rex::ConnectionRefused, Rex::ConnectionError
+      return false
+    end
+    sock.close
+    return true
+  end
+
+  def do_grab_domain_ip_history(hostname, proxies)
+    begin
+      cli = Rex::Proto::Http::Client.new('www.prepostseo.com', 443, {}, true, nil, proxies)
+      cli.connect
+
```

Comment:
Good use of the namespace you have, but as a language construct we use mixins for this. _Exploit::Remote::Http_, I believe is what would be included to abstract this part. Take look at some other http scanners to see how the convenience methods handle this and underlying mechanisms like socket registration.

Reply:
Thank you, I have already made this remark. in fact it's my second code Ruby and I must admit that it's a bit confusing.

I did not find a use case in this context w/ scanner then I relied on modules/auxiliary/gather/censys_search.rb